src/kdf.c

```diff
@@ -124,49 +241,245 @@ static int p11prov_hkdf_derive(void *ctx, unsigned char *key, size_t keylen,
+    i += sizeof(uint16_t);
+    info[i] = (uint8_t)(prefixlen + labellen);
+    i += 1;
+    memcpy(&info[i], prefix, prefixlen);
+    i += prefixlen;
+    memcpy(&info[i], label, labellen);
+    i += labellen;
+    info[i] = (uint8_t)(datalen);
+    i += 1;
+    if (datalen > 0) {
+        memcpy(&info[i], data, datalen);
+        i += datalen;
     }
+    if (infolen != i) {
+        OPENSSL_cleanse(&info, i);
+        return CKR_HOST_MEMORY;
+    }
+
+    params.pInfo = info;
+    params.ulInfoLen = infolen;
 
-    ret = p11prov_derive_key(hkdfctx->provctx, slotid, &mechanism, pkey_handle,
-                             key_template, 5, &hkdfctx->session, &dkey_handle);
+    ret = inner_derive_key(hkdfctx->provctx, keyobj, &hkdfctx->session,
+                           &mechanism, keylen, dkey_handle);
+
+    OPENSSL_cleanse(&info, infolen);
+    return ret;
+}
+
+static CK_RV p11prov_tls13_derive_secret(P11PROV_KDF_CTX *hkdfctx,
+                                         P11PROV_OBJ *keyobj, size_t keylen,
+                                         CK_OBJECT_HANDLE *dkey_handle)
+{
+    P11PROV_OBJ *zerokey = NULL;
+    CK_HKDF_PARAMS params = {
+        .bExtract = CK_TRUE,
+        .bExpand = CK_FALSE,
+        .prfHashMechanism = hkdfctx->hash_mech,
+        .ulSaltType = CKF_HKDF_SALT_DATA,
+        .hSaltKey = CK_INVALID_HANDLE,
+        .pInfo = NULL,
+        .ulInfoLen = 0,
+    };
+    CK_MECHANISM mechanism = {
+        .mechanism = CKM_HKDF_DATA,
+        .pParameter = &params,
+        .ulParameterLen = sizeof(params),
+    };
+    uint8_t saltbuf[EVP_MAX_MD_SIZE] = { 0 };
+    uint8_t zerobuf[EVP_MAX_MD_SIZE] = { 0 };
+    size_t saltlen;
+    size_t hashlen;
+    CK_RV ret;
+
+    ret = p11prov_digest_get_digest_size(hkdfctx->hash_mech, &hashlen);
     if (ret != CKR_OK) {
+        return ret;
+    }
+    saltlen = hashlen;
+
+    if (hkdfctx->salt) {
+        P11PROV_OBJ *ek = NULL;
+        unsigned char info[hashlen];
+        const char *mdname;
+        data_buffer digest_data[1] = { 0 }; /* intentionally empty */
+        data_buffer digest = { .data = info, .length = hashlen };
+        CK_OBJECT_HANDLE skey_handle;
+
+        /* OpenSSL special cases this in an odd way and regenerates a hash as
+         * if an empty message was received. */
+        ret = p11prov_digest_get_name(hkdfctx->hash_mech, &mdname);
+        if (ret != CKR_OK) {
+            return ret;
+        }
+
+        ret = p11prov_digest_util(hkdfctx->provctx, mdname, NULL, digest_data,
+                                  &digest);
+        if (ret != CKR_OK) {
+            return ret;
+        }
+
+        /* In OpenSSL the salt is used as the derivation key */
+        /* FIXME: OpenSSL sets the length to hashlen w/o checks,
+         * this may be a bug or intentional, further investigation needed
+         * [opened https://github.com/openssl/openssl/issues/25557] */
+        ret = inner_pkcs11_key(hkdfctx, hkdfctx->salt, hkdfctx->saltlen, &ek);
```

Comment:
IIUC, in this function we are creating an object in Token also, but at line 398, we are just removing the object from p11prov pool and not removing it from token?